### ISSUE TITLE
cycle error building rustc with debug logging

### ISSUE THREAD

**Issue by ehuss** [rust-lang/rust#61577]
When using `RUSTC_LOG` to debug building rustc itself, I ran into this error:

```
error[E0391]: cycle detected when processing `hir::def::<impl at src/librustc/hir/def.rs:255:1: 270:2>::present_items::{{opaque}}#0`
   --> src/librustc/hir/def.rs:262:35
    |
262 |     pub fn present_items(self) -> impl Iterator<Item=T> {
    |                                   ^^^^^^^^^^^^^^^^^^^^^
    |                                   |
    |                                   in this expansion of `desugaring of `existential type``
    |                                   in this macro invocation
    |
note: ...which requires processing `hir::def::<impl at src/librustc/hir/def.rs:255:1: 270:2>::present_items::{{opaque}}#0`...
   --> src/librustc/hir/def.rs:262:35
    |
262 |     pub fn present_items(self) -> impl Iterator<Item=T> {
    |                                   ^^^^^^^^^^^^^^^^^^^^^
    |                                   |
    |                                   in this expansion of `desugaring of `existential type``
    |                                   in this macro invocation
note: ...which requires processing `hir::def::<impl at src/librustc/hir/def.rs:255:1: 270:2>::present_items::{{opaque}}#0`...
   --> src/librustc/hir/def.rs:262:35
    |
262 |     pub fn present_items(self) -> impl Iterator<Item=T> {
    |                                   ^^^^^^^^^^^^^^^^^^^^^
    |                                   |
    |                                   in this expansion of `desugaring of `existential type``
    |                                   in this macro invocation
    = note: ...which again requires processing `hir::def::<impl at src/librustc/hir/def.rs:255:1: 270:2>::present_items::{{opaque}}#0`, completing the cycle
note: cycle used when processing `hir::def::<impl at src/librustc/hir/def.rs:255:1: 270:2>::present_items`
   --> src/librustc/hir/def.rs:262:5
    |
262 |     pub fn present_items(self) -> impl Iterator<Item=T> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Repro:
1. Set `debug = true` in `config.toml`.
2. `./x.py build` to get a debug compiler (and go through several steps in stage 1 we want to skip).
3. `touch src/librustc/lib.rs`
4. `RUSTC_LOG=rustc::ty=debug ./x.py build --stage=1 --keep-stage=0 > dlog 2>&1`

The last step should rebuild librustc using a compiler with debugging enabled.  I imagine there is a faster way to do this, but these are the simplest instructions I could think of.

Is this known/expected?  I have no idea what the error means, but it makes debugging more difficult.

Currently on master (7cdaffd7962c4aae0cadd82baa241901b03f9458).

**Comment by jonas-schievink** [comment]
Ran into the same problem while investigating https://github.com/rust-lang/rust/issues/60414#issuecomment-488021357, very annoying indeed!

**Comment by Aaron1011** [comment]
I ran into a similar issue when using `RUSTC_LOG=rustc::ty=debug` - however, I'm not sure if this is exactly the same cause.

I tracked down the cycle error I saw to [the pretty-printing code for opaque types](https://github.com/rust-lang/rust/blob/618768492f0c731fcb770dc2d178abe840846419/src/librustc/ty/print/pretty.rs#L580). When not printing in 'verbose mode', it attempts to call `tcx.predicates_of` on the opaque type. If we're already trying to get the predicates of an opaque type, this results in a cycle error.

As a hacky workaround, I changed `if self.tcx().sess.verbose()` to `if self.tcx().sess.verbose() || true`. However, the real fix is to avoid running queries during type pretty printing.
